nemo_aligner/utils/trt_llm.py

```diff
@@ -1,136 +1,111 @@
-                nemo_model=model, nemo_model_config=self.cfg,
+                nemo_model = model, 
+                nemo_model_config = self.cfg, 
             )
 
-    def generate(self, inputs):
-        stop_words = self.stop_words
-        output_ids = self.forward(inputs, stop_words)
-
-        if output_ids is not None:
-            mbs = output_ids.shape[0]
-            if mbs == 1:
-                output_ids = output_ids.view([1, output_ids.shape[-1]])
-            else:
-                output_ids = output_ids.squeeze()
-            output_ids = output_ids.to(torch.int64)
-
-        output_ids = broadcast_2d_tensor_within_mp(output_ids, dtype=torch.long)
-        print("### RANK", torch.distributed.get_rank(), output_ids)
-
-        sentences = [self.tokenizer.ids_to_text(output.tolist()) for output in output_ids]
-        output_ids = output_ids.tolist()
-
-        output = {
-            "token_ids": output_ids,
-            "sentences": sentences,
-        }
 
-        return output
+    def generate(self, inputs):
+        prompt_tokens, prompt_lengths = inputs
 
-    def forward(self, inputs, stop_ids):
-        from nemo.export.trt_llm.tensorrt_llm_run import tensorrt_llm_worker_context
+        batch_input_ids = []
+        for idx in range(prompt_tokens.shape[0]):
+            batch_input_ids.append(prompt_tokens[idx][0:prompt_lengths[idx]].cpu())
 
-        decoder = tensorrt_llm_worker_context.decoder
-        from tensorrt_llm.runtime import ModelRunner
+        output_dict = self.trt_llm_exporter.model_runner.generate(
+            batch_input_ids=batch_input_ids,
+            sampling_config=self.sampling_config,
+            streaming=False)
 
-        self.model_runner = ModelRunner(
-            session=decoder,
-            max_batch_size=self.generation_batch_size,
-            max_input_len=self.max_context_length,
-            max_seq_len=4096,  # TODO: fix
-            max_beam_width=1,
-        )
+        output_ids = output_dict['output_ids']
+        resp_lens = output_dict['sequence_lengths'].squeeze()
 
-        prompt_tokens, prompt_lengths = inputs
+        # remove beam dim from output_ids: [mbs, beam_dim, sequence len]
+        mbs = output_ids.shape[0]
+        if mbs == 1:
+            output_ids = output_ids.view([1,output_ids.shape[-1]])
+        else:
+            output_ids = output_ids.squeeze()
+        output_ids = output_ids.to(torch.int64)
 
-        batch_input_ids = []
-        for idx in range(prompt_tokens.shape[0]):
-            batch_input_ids.append(prompt_tokens[idx][0 : prompt_lengths[idx]].cpu())
+        #broadcast output to all PP ranks
+        if parallel_state.get_pipeline_model_parallel_world_size() > 1 and not self.reshard_model:  
+            group = parallel_state.get_pipeline_model_parallel_group()
+            src = parallel_state.get_pipeline_model_parallel_first_rank()
+            output_ids = broadcast_2d_tensor(output_ids, src, group, dtype=output_ids.dtype)
```

Comment:
https://github.com/NVIDIA/NeMo-Aligner/blob/geshen/main_trt/nemo_aligner/utils/distributed.py#L102

I think you can use this function instead. The groups are unchanged when we don't have resharding and when there's resharding this should be fairly trivial(though if you think we have perf hit lmk)

Reply:
wouldnt this also broadcast to gpus in the same TP group? Which TRTLLM already has equal across TP

Reply:
yeah do you think that's too much of a perf hit or doesn't matter?

Reply:
I dont think it would affect perf but it does make the code less clear
This broadcast only happens without resharding, do we need to guard it?

Reply:
okay let's guard on no reshard and then use that function then?